### PR TITLE
Turn titan into read-only on error during gc and version_change

### PR DESCRIPTION
**Pull request by wujy-cs**
- Set background error if encountered a critical error. Now it's set for two cases
    - BackgroundGC() error
    - LogAndApply() error after generate new blob file
- If background error is set, all following foreground write operations will be refused and return background error status

### CODE REVIEW

File: src/db_impl_gc.cc [hunk 1]
```diff
@@ -108,13 +107,33 @@ Status TitanDBImpl::BackgroundGC(LogBuffer* log_buffer) {
...
   return s;
 }
 
+Status TitanDBImpl::SetBGError(const Status& s) {
+  if (s.ok()) return s;
+  mutex_.AssertHeld();
+  Status bg_err = s;
+  if (!db_options_.listeners.empty()) {
+    // TODO(@jiayu) : check if mutex_ is freeable for futrue use case
+    mutex_.Unlock();
```
[the remark this answers is not in the document]
Comment (wujy-cs) [reply src/db_impl_gc.cc]: bg_err is a local variable, won't be accessed concurrently.

File: src/titan_db_test.cc [hunk 2]
```diff
@@ -748,6 +814,9 @@ TEST_F(TitanDBTest, FallbackModeEncounterMissingBlobFile) {
   VerifyDB({{"bar", "v1"}});
 }
 
+TEST_F(TitanDBTest, BackgroundErrorHandling) { BackgroundErrorHandling(); }
```
[the remark this answers is not in the document]
Comment (wujy-cs) [reply src/titan_db_test.cc]: ok

File: src/db_impl_gc.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (wujy-cs) [reply src/db_impl_gc.cc]: I'm not sure if listeners would be modified, it seems no API to do that unless reopen db

File: src/titan_db_test.cc [hunk 3]
```diff
@@ -748,6 +772,48 @@ TEST_F(TitanDBTest, FallbackModeEncounterMissingBlobFile) {
...
 
+TEST_F(TitanDBTest, BackgroundErrorHandling) {
+  options_.listeners.emplace_back(std::make_shared<BGErrorListener>());
+  Open();
+  std::string key = "key", val;
+  SetBGError(Status::IOError(""));
+  // BG error is restored by listener for first time
+  ASSERT_OK(db_->Put(WriteOptions(), key, val));
+  SetBGError(Status::IOError(""));
+  ASSERT_OK(db_->Get(ReadOptions(), key, &val));
```
Comment (yiwu-arbug) [inline src/titan_db_test.cc]: Put a non-empty value and verify the value return from `Get`?

**Comment by yiwu-arbug** [comment]
@tabokie thanks for reviewing btw :)

**Comment by wujy-cs** [comment]
@yiwu-arbug @tabokie All modified, thanks for review.

**Pull request merged**